synapse/lib/cmdr.py

```diff
@@ -9,36 +20,595 @@
+
+    Examples:
+
+        Open a connection to a server and drop into an interactive command loop:
+
+            cliprox = openurl('tcp://cliserver.com:8675/mycli')
+            cliprox.runCmdLoop()
+
+    Returns:
+        CliProxy: A CliProxy object which can be used to interact with a remote object.
+    '''
+    cliserver = s_telepath.openurl(url, **opts)
+    return CliProxy(cliserver)
 
 def getItemCmdr(item, outp=None, **opts):
     '''
     Construct and return a cmdr for the given item.
 
-    Example:
+    Args:
+        item (object): Object being commanded
+        outp (s_output.Output): An Output object for printing data with.
+        **opts: Additional options passed to the Cli.
 
-        cmdr = getItemCmdr(foo)
+    Examples:
+        Get the CLI for an object and run the command loop:
 
-    '''
-    cmdr = s_cli.Cli(item, outp=outp)
-
-    refl = s_reflect.getItemInfo(item)
-    if refl is None:
-        return cmdr
+            cmdr = getItemCmdr(foo)
+            cmdr.runCmdLoop()
 
-    for name in refl.get('inherits', ()):
-        for mixi in s_mixins.getSynMixins('cmdr', name):
-            cmdr.addCmdClass(mixi)
+    Returns:
+        s_cli.Cli: A Cli for the object.
+    '''
+    cmdr = s_cli.Cli(item, outp=outp, **opts)
+    cmdr.reflectItem()
 
     return cmdr
 
 def runItemCmdr(item, outp=None, **opts):
     '''
-    Create a cmdr for the given item and run the cmd loop.
+    Create a cmdr for the given item and run the cmd loop directly.
 
-    Example:
+    Args:
+        item (object): Object being commanded
+        outp (s_output.Output): An Output object for printing data with.
+        **opts: Additional options passed to the Cli.
 
-        runItemCmdr(foo)
+    Examples:
+        Get and run the cmdloop for an object directly:
 
+            runItemCmdr(foo)
+
+    Returns:
+        None
     '''
     cmdr = getItemCmdr(item, outp=outp, **opts)
     cmdr.runCmdLoop()
+
+class CliClient(s_cli.Cli):
+    '''
+    The CliClient is used to implement the server-side Cli object.
+
+    This is created and torn down per user.
+
+    Third party implementations may override this class and provide an updated
+    python path to their object as a configable option to the CliServer as
+    ``cliserver:clientctor``.
+    '''
+    def runCmdLoop(self):
+        '''
+        This is a invalid function to execute on the server side.
+
+        Raises:
+            NoSuchImpl: When called.
+        '''
+        raise s_common.NoSuchImpl(mesg='CliClient cannot run a user cmdloop')
+
+    def runCmdLine(self, line):
+        '''
+        Run a single command line.
+
+        Args:
+            line (str): Line to execute.
+
+        Examples:
+            Execute the 'woot' command with the 'help' switch:
+
+                cli.runCmdLine('woot --help')
+
+        Returns:
+            object: Arbitrary data from the cmd class.
+        '''
+        ret = None
+
+        name = line.split(None, 1)[0]
```

Comment:
Should we check that line is a non-empty string here? It looks like we're checking it before the call to this fn. Could we move it in here?